csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll

```diff
@@ -0,0 +1,998 @@
+    Type resultType, boolean isLValue) {
+    tag = OnlyInstructionTag() and
+    opcode instanceof Opcode::NoOp and
+    resultType instanceof VoidType and
+    isLValue = false
+  }
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    none()
+  }
+  
+  override Instruction getChildSuccessor(TranslatedElement child) {
+    none()
+  }
+}
+
+class TranslatedBreakStmt extends TranslatedSpecificJump {
+  override BreakStmt stmt;
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    tag = OnlyInstructionTag() and
+    kind instanceof GotoEdge and
+    result = getEnclosingLoopOrSwitchNextInstr(stmt)
+  }
+  
+  private Instruction getEnclosingLoopOrSwitchNextInstr(Stmt crtStmt) {
+    if (crtStmt instanceof LoopStmt or crtStmt instanceof SwitchStmt) then
+      result = getTranslatedStmt(crtStmt).getParent().getChildSuccessor(getTranslatedStmt(crtStmt))
+    else
+      result = getEnclosingLoopOrSwitchNextInstr(crtStmt.getParent())
+  }
+}
+
+class TranslatedGotoLabelStmt extends TranslatedSpecificJump {
+  override GotoLabelStmt stmt;
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    tag = OnlyInstructionTag() and
+    kind instanceof GotoEdge and
+    result = getTranslatedStmt(stmt.getTarget()).getFirstInstruction()
+  }
+}
+
+class TranslatedGotoCaseStmt extends TranslatedSpecificJump {
+  override GotoCaseStmt stmt;
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    tag = OnlyInstructionTag() and
+    kind instanceof GotoEdge and
+    result = getCase(stmt, stmt.getExpr()).getFirstInstruction()
+  }
+  
+  private TranslatedStmt getCase(Stmt crtStmt, Expr expr) {
+    if (crtStmt instanceof SwitchStmt) then
+      exists (CaseStmt caseStmt |
+        caseStmt = crtStmt.(SwitchStmt).getACase() and
+        // We check for the constant value of the expression
+        // since we can't check for equality between `PatternExpr` and `Expr`
+        caseStmt.getPattern().getValue() = expr.getValue() and
+        result = getTranslatedStmt(caseStmt)
+      )
+    else
+      result = getCase(crtStmt.getParent(), expr)
+  }
+}
+
+class TranslatedGotoDefaultStmt extends TranslatedSpecificJump {
+  override GotoDefaultStmt stmt;
+
+  override Instruction getInstructionSuccessor(InstructionTag tag,
+    EdgeKind kind) {
+    tag = OnlyInstructionTag() and
+    kind instanceof GotoEdge and
+    result = getDefaultCase(stmt).getFirstInstruction()
+  }
+  
+  private TranslatedStmt getDefaultCase(Stmt crtStmt) {
+    if (crtStmt instanceof SwitchStmt) then
+      exists (CaseStmt caseStmt |
+        caseStmt = crtStmt.(SwitchStmt).getDefaultCase() and
+        result = getTranslatedStmt(caseStmt)
+      )
+    else
+      result = getDefaultCase(crtStmt.getParent())
+  }
+}
+
+class TranslatedSwitchStmt extends TranslatedStmt {
```

Comment:
Arguably, we could model this as a chain of ifs (unlike C++), so you could synthesise  something like
```
switch (x)
{
case 1:
case "foo:
  ...
  break;
case 2:
  ...
}
```
as
```
tmp = x;
if (tmp == 1 || tmp == "foo")
   ...
else if (tmp == 2)
  ...
```

Reply:
Also have `when` conditions to consider.

Reply:
Should we do this for this PR? I believe this is a good idea, since treating `when` and pattern matching would be easier inside conditions.

Reply:
Do it later.